File: leetcode_oj/count_primes.py
'''
Description:

Count the number of prime numbers less than a non-negative number, n
'''
class Solution:
    # @param {integer} n
    # @return {integer}
    def countPrimes(self, n):
		end = n
		c = 1
		nums = {}
		for j in range(2, end, 2):
			try:
				nums[j] = 1 
			except:
				pass
		
		for n in range(3, end):
			if n in nums:
				continue
			for i in range(2, n):
				if (n % i) == 0:
					break
			if i == (n-1):
				c += 1
				# Add multiple of n
				for j in range(n, end, n):
					try:
						nums[j] = 1 
					except:
						pass
		return c

s = Solution()
print s.countPrimes(1000)

File: leetcode_oj/countPrimes.cpp
class Solution {
public:
    int countPrimes(int n) {
        vector<bool> isprimes(max(n,2), true);
        isprimes[0] = false; isprimes[1] = false;
        int x = 2;
        while (x * x < n) {
            if (isprimes[x]) {
                int p = x * x;
                while (p < n) {
                    isprimes[p] = false;
                    p += x;
                }
            }
            x += 1;
        }
        return count_if(isprimes.begin(), isprimes.end(), [](bool i){return i==true;});
    }
};
